src/Tools/BuildValidator/LocalSourceResolver.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+
+namespace BuildValidator
+{
+    /// <summary>
+    /// Roslyn specific implementation for looking for files
+    /// in the Roslyn repo
+    /// </summary>
+    internal class LocalSourceResolver : ISourceResolver
+    {
+        private readonly DirectoryInfo _baseDirectory;
+        private readonly ILogger _logger;
+
+        public LocalSourceResolver(ILoggerFactory loggerFactory)
+        {
+            _baseDirectory = GetSourceDirectory();
+            _logger = loggerFactory.CreateLogger<LocalSourceResolver>();
+
+            _logger.LogInformation($"Source Base Directory: {_baseDirectory}");
+        }
+
+        public Task<SourceText> ResolveSourceAsync(string name, Encoding encoding)
+        {
+            if (!File.Exists(name))
```

Comment:
Naughty naughty 😄

Reply:
While I know the complaint with `File.Exists`, is there a better way to do path checking? This is used to add the base directory when needed. Do I just try/catch the read, and try again with the new path?

Reply:
Given the way the debugger works the only sensible thing to do is use `File.Exists`, then read to avoid annoying first chance exceptions. I was mostly joking here. Think this approach is fine.